packages/serverless/src/domain/use-case/collection.ts

```diff
@@ -71,37 +83,85 @@ async function createCollection(
-    collectionName
-  );
-  const indexes = await modelCollection.listIndexes();
-  const sizeOnDisk = await modelCollection.size();
+  if (
+    skipPermissionCheck ||
+    (await hasCollectionPermission(databaseName, collectionName, actor, 'read'))
+  ) {
+    const modelCollection = ModelCollection.getInstance(
+      databaseName,
+      collectionName
+    );
+    const indexes = await modelCollection.listIndexes();
+    const sizeOnDisk = await modelCollection.size();
 
-  const schema = await getSchemaDefinition(databaseName, collectionName);
-  const ownership = await readMetadata(
-    databaseName,
-    collectionName,
-    null,
-    ZKDATABASE_USER_NOBODY
-  );
+    const schema = await getSchemaDefinition(
+      databaseName,
+      collectionName,
+      actor,
+      true
+    );
+
+    const ownership = await readMetadata(
+      databaseName,
+      collectionName,
+      null,
+      actor
+    );
 
-  await ModelCollection.getInstance(databaseName, collectionName).size();
+    await ModelCollection.getInstance(databaseName, collectionName).size();
 
-  return { name: collectionName, indexes, schema, ownership, sizeOnDisk };
+    return { name: collectionName, indexes, schema, ownership, sizeOnDisk };
+  }
+
+  throw new Error(
+    `Access denied: Actor '${actor}' does not have 'read' permission for collection '${collectionName}'.`
+  );
 }
 
-async function listCollections(databaseName: string): Promise<Collection[]> {
-  const collectionNames =
-    await ModelDatabase.getInstance(databaseName).listCollections();
+async function listCollections(
+  databaseName: string,
+  actor: string
+): Promise<Collection[]> {
+  let availableCollections: string[] = [];
+
+  if (await isDatabaseOwner(databaseName, actor)) {
+    availableCollections =
+      await ModelDatabase.getInstance(databaseName).listCollections();
+  } else {
+    const collectionsMetadata = await (
+      await ModelCollectionMetadata.getInstance(databaseName).find()
+    ).toArray();
+
+    const modelUserGroup = new ModelUserGroup(databaseName);
+
+    const actorGroups = await modelUserGroup.listGroupByUserName(actor);
+
+    for (const metadata of collectionsMetadata) {
+      if (
+        (metadata.owner === actor &&
+          PermissionBinary.fromBinaryPermission(metadata.permissionOwner)[
+            'read'
+          ]) ||
+        (actorGroups.includes(metadata.group) &&
+          PermissionBinary.fromBinaryPermission(metadata.permissionGroup)[
+            'read'
+          ]) ||
+        PermissionBinary.fromBinary(metadata.permissionOther)['read']
+      ) {
+        availableCollections.push(metadata.collection);
+      }
+    }
+  }
 
   return (
     await Fill(
-      collectionNames.map(
+      availableCollections.map(
         (collectionName) => async () =>
-          readCollectionInfo(databaseName, collectionName)
+          readCollectionInfo(databaseName, collectionName, actor, false)
       )
     )
   ).map(({ result }) => result);
```

Comment:
we need to filter out which one is null as well, because it can be null and when we return the graphql will throw error null mismatch with that result type
```ts
.filter(Boolean)
```